In [ ]:
import os

os.makedirs("/content/results", exist_ok=True)


In [ ]:
!python "posthog copy.py"


📅 Semana actual:
   Inicio: 2025-11-24 00:00:00
   Fin: 2025-11-30 23:59:59

📊 Consultando eventos...

✓ Eventos guardados: /content/results/posthog_events_2025_12_01.csv
  Total registros: 1623

📊 Consultando sesiones...

✓ Sesiones guardadas: /content/results/posthog_sessions_2025_12_01.csv
  Total registros: 106

✓ Exportación completada


# Task
Load `/content/results/posthog_events_2025_12_01.csv` into a DataFrame named `df_events` and `/content/results/posthog_sessions_2025_12_01.csv` into a DataFrame named `df_sessions`. Then, convert the 'timestamp' column in `df_events` and the 'session_start' and 'session_end' columns in `df_sessions` to datetime objects.

## Convert Time Columns to Datetime

### Subtask:
Convert all relevant time-related columns in `df_events` and `df_sessions` (e.g., 'timestamp', 'session_start', 'session_end') to datetime objects to ensure proper time-based operations.


**Reasoning**:
To begin, I will import the pandas library, load both the events and sessions CSV files into DataFrames, and then convert all specified time-related columns ('timestamp', 'session_start', 'session_end') to datetime objects to ensure they are in the correct format for subsequent time-based analysis.



In [ ]:
import pandas as pd

# Load the events data
df_events = pd.read_csv('/content/results/posthog_events_2025_12_01.csv')

# Load the sessions data
df_sessions = pd.read_csv('/content/results/posthog_sessions_2025_12_01.csv')

# Convert 'timestamp' column in df_events to datetime objects
df_events['timestamp'] = pd.to_datetime(df_events['timestamp'])

# Convert 'session_start' and 'session_end' columns in df_sessions to datetime objects
df_sessions['session_start'] = pd.to_datetime(df_sessions['session_start'])
df_sessions['session_end'] = pd.to_datetime(df_sessions['session_end'])

print("Time columns successfully converted to datetime objects in both df_events and df_sessions.")
print("df_events info:")
df_events.info()
print("\ndf_sessions info:")
df_sessions.info()

ValueError: time data "2025-11-25T12:25:58Z" doesn't match format "%Y-%m-%dT%H:%M:%S.%f%z", at position 1396. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.

**Reasoning**:
The previous code failed because the `pd.to_datetime` function could not infer the correct format for the 'timestamp' column in `df_events` and the 'session_start' and 'session_end' columns in `df_sessions`. I will re-run the conversion, explicitly specifying `format='ISO8601'` for `df_events['timestamp']` and `format='%d-%m-%Y %H:%M:%S'` for `df_sessions['session_start']` and `df_sessions['session_end']` to ensure correct parsing.



In [ ]:
import pandas as pd

# Load the events data
df_events = pd.read_csv('/content/results/posthog_events_2025_12_01.csv')

# Load the sessions data
df_sessions = pd.read_csv('/content/results/posthog_sessions_2025_12_01.csv')

# Convert 'timestamp' column in df_events to datetime objects, specifying ISO8601 format
df_events['timestamp'] = pd.to_datetime(df_events['timestamp'], format='ISO8601')

# Convert 'session_start' and 'session_end' columns in df_sessions to datetime objects, specifying the format
df_sessions['session_start'] = pd.to_datetime(df_sessions['session_start'], format='%d-%m-%Y %H:%M:%S')
df_sessions['session_end'] = pd.to_datetime(df_sessions['session_end'], format='%d-%m-%Y %H:%M:%S')

print("Time columns successfully converted to datetime objects in both df_events and df_sessions.")
print("df_events info:")
df_events.info()
print("\ndf_sessions info:")
df_sessions.info()

KeyError: 'session_start'

**Reasoning**:
The previous code failed because the column names 'session_start' and 'session_end' were not found in `df_sessions`. I have inspected the kernel state and identified that the correct column names are 'start' and 'end' in `df_sessions`. I will modify the code to use these correct column names for conversion to datetime objects, while keeping the explicit format for both `df_events['timestamp']` and `df_sessions['start']` and `df_sessions['end']`.



In [ ]:
import pandas as pd

# Load the events data
df_events = pd.read_csv('/content/results/posthog_events_2025_12_01.csv')

# Load the sessions data
df_sessions = pd.read_csv('/content/results/posthog_sessions_2025_12_01.csv')

# Convert 'timestamp' column in df_events to datetime objects, specifying ISO8601 format
df_events['timestamp'] = pd.to_datetime(df_events['timestamp'], format='ISO8601')

# Convert 'start' and 'end' columns in df_sessions to datetime objects, specifying the format
df_sessions['start'] = pd.to_datetime(df_sessions['start'], format='%d-%m-%Y %H:%M:%S')
df_sessions['end'] = pd.to_datetime(df_sessions['end'], format='%d-%m-%Y %H:%M:%S')

print("Time columns successfully converted to datetime objects in both df_events and df_sessions.")
print("df_events info:")
df_events.info()
print("\ndf_sessions info:")
df_sessions.info()

Time columns successfully converted to datetime objects in both df_events and df_sessions.
df_events info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1623 entries, 0 to 1622
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype              
---  ------       --------------  -----              
 0   distinct_id  1623 non-null   object             
 1   event        1623 non-null   object             
 2   screen       1506 non-null   object             
 3   sub_screen   181 non-null    object             
 4   action       290 non-null    object             
 5   datetime     1623 non-null   object             
 6   timestamp    1623 non-null   datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1), object(6)
memory usage: 88.9+ KB

df_sessions info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 106 entries, 0 to 105
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   

## Merge Events and Sessions

### Subtask:
Merge `df_events` and `df_sessions` based on `distinct_id` and then filter events to ensure their `timestamp` falls within the `session_start` and `session_end` of the corresponding session. This step will prepare the data for feature calculation by associating events with their respective sessions.


**Reasoning**:
To associate events with their respective sessions, I will first convert the 'start' and 'end' columns in `df_sessions` to UTC to match the timezone of `df_events['timestamp']`. Then, I will merge `df_events` with `df_sessions` using a left join on 'distinct_id'. Finally, I will filter the merged DataFrame to include only events whose 'timestamp' falls within the 'start' and 'end' times of their associated session.



In [ ]:
import pandas as pd

# Ensure df_sessions 'start' and 'end' columns are timezone-aware and in UTC
df_sessions['start'] = df_sessions['start'].dt.tz_localize('UTC')
df_sessions['end'] = df_sessions['end'].dt.tz_localize('UTC')

# Merge df_events and df_sessions on 'distinct_id'
df_merged = pd.merge(df_events, df_sessions, on='distinct_id', how='left')

# Filter events where timestamp is within the session start and end times
df_merged_filtered = df_merged[
    (df_merged['timestamp'] >= df_merged['start']) &
    (df_merged['timestamp'] <= df_merged['end'])
].copy()

print("Merged and filtered DataFrame info:")
df_merged_filtered.info()
print("\nFirst 5 rows of the merged and filtered DataFrame:")
print(df_merged_filtered.head())

Merged and filtered DataFrame info:
<class 'pandas.core.frame.DataFrame'>
Index: 29 entries, 4116 to 9618
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype              
---  ------            --------------  -----              
 0   distinct_id       29 non-null     object             
 1   event             29 non-null     object             
 2   screen            23 non-null     object             
 3   sub_screen        9 non-null      object             
 4   action            0 non-null      object             
 5   datetime          29 non-null     object             
 6   timestamp         29 non-null     datetime64[ns, UTC]
 7   start             29 non-null     datetime64[ns, UTC]
 8   end               29 non-null     datetime64[ns, UTC]
 9   max_insert        29 non-null     object             
 10  duration          29 non-null     float64            
 11  $start_timestamp  29 non-null     object             
dtypes: datetime64[ns, UTC](3), flo

## Calculate Session Metrics

### Subtask:
Create the `df_features` DataFrame. For each unique session, calculate the following metrics: count of `marker_select` events, count of `search` or `filter_select` events, count of `page_view` events, a binary flag (1 or 0) indicating if any `download` event occurred, and the 'real duration' of the session, calculated as the difference between the timestamp of the last event and the first event within that session.


**Reasoning**:
I need to calculate session-level metrics for `df_features`. This requires grouping `df_merged_filtered` by `distinct_id` and `start`, then applying aggregation functions to count specific events, check for download events, and compute the real duration of each session.



In [ ]:
import pandas as pd

# Group by distinct_id and session start time ('start') and calculate features
df_features = df_merged_filtered.groupby(['distinct_id', 'start']).agg(
    marker_select_count=('event', lambda x: (x == 'marker_select').sum()),
    search_filter_select_count=('event', lambda x: ((x == 'search') | (x == 'filter_select')).sum()),
    page_view_count=('event', lambda x: (x == 'page_view').sum()),
    download_flag=('event', lambda x: (x == 'download').any().astype(int)),
    real_duration=('timestamp', lambda x: (x.max() - x.min()).total_seconds())
).reset_index()

print("df_features created with session metrics:")
df_features.info()
print("\nFirst 5 rows of df_features:")
print(df_features.head())

df_features created with session metrics:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   distinct_id                 4 non-null      object             
 1   start                       4 non-null      datetime64[ns, UTC]
 2   marker_select_count         4 non-null      int64              
 3   search_filter_select_count  4 non-null      int64              
 4   page_view_count             4 non-null      int64              
 5   download_flag               4 non-null      int64              
 6   real_duration               4 non-null      float64            
dtypes: datetime64[ns, UTC](1), float64(1), int64(4), object(1)
memory usage: 356.0+ bytes

First 5 rows of df_features:
                               distinct_id                     start  \
0  antonietta.borzacchiello@deceuninck.com 2

## Handle Missing Values and Scale Features

### Subtask:
Fill any `NaN` values in `df_features` with 0. Then, scale the numerical features in `df_features` using a suitable scaler (e.g., StandardScaler) to prepare the data for clustering.


**Reasoning**:
First, I will identify the numerical columns in `df_features` to be scaled. Then, I will fill any potential `NaN` values in these columns with 0, although `df_features` currently shows no missing values. Finally, I will import `StandardScaler` from `sklearn.preprocessing`, initialize it, and apply it to the identified numerical columns, storing the scaled data in a new DataFrame.



In [ ]:
from sklearn.preprocessing import StandardScaler
import pandas as pd

# Identify numerical columns to scale
numerical_cols = [
    'marker_select_count',
    'search_filter_select_count',
    'page_view_count',
    'download_flag',
    'real_duration'
]

# Fill any NaN values in numerical columns with 0 (as per instruction, even if currently no NaNs)
df_features[numerical_cols] = df_features[numerical_cols].fillna(0)

# Initialize StandardScaler
scaler = StandardScaler()

# Apply scaler to numerical columns and create a new DataFrame for scaled features
df_scaled_features = pd.DataFrame(scaler.fit_transform(df_features[numerical_cols]),
                                  columns=numerical_cols,
                                  index=df_features.index)

# Optionally, add back the non-numerical columns for context
df_scaled_features[['distinct_id', 'start']] = df_features[['distinct_id', 'start']]

print("Numerical features in df_features scaled and stored in df_scaled_features.")
print("df_scaled_features info:")
df_scaled_features.info()
print("\nFirst 5 rows of df_scaled_features:")
print(df_scaled_features.head())

Numerical features in df_features scaled and stored in df_scaled_features.
df_scaled_features info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 7 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   marker_select_count         4 non-null      float64            
 1   search_filter_select_count  4 non-null      float64            
 2   page_view_count             4 non-null      float64            
 3   download_flag               4 non-null      float64            
 4   real_duration               4 non-null      float64            
 5   distinct_id                 4 non-null      object             
 6   start                       4 non-null      datetime64[ns, UTC]
dtypes: datetime64[ns, UTC](1), float64(5), object(1)
memory usage: 356.0+ bytes

First 5 rows of df_scaled_features:
   marker_select_count  search_filter_select_count  page_view

## Perform KMeans Clustering

### Subtask:
Apply the KMeans clustering algorithm with `n_clusters=3` to the scaled `df_features` data. Assign the resulting cluster labels back to the `df_features` DataFrame.


**Reasoning**:
To perform KMeans clustering, I will import `KMeans` from `sklearn.cluster`, instantiate the model with the specified parameters, fit it to the numerical scaled features, and then assign the resulting cluster labels to a new 'cluster' column in `df_features`.



In [ ]:
from sklearn.cluster import KMeans

# Instantiate KMeans with n_clusters=3, random_state=42, and n_init='auto'
kmeans = KMeans(n_clusters=3, random_state=42, n_init='auto')

# Fit the KMeans model to the numerical columns of df_scaled_features
kmeans.fit(df_scaled_features[numerical_cols])

# Assign the cluster labels back to the original df_features DataFrame
df_features['cluster'] = kmeans.labels_

print("df_features with assigned cluster labels:")
print(df_features.head())

df_features with assigned cluster labels:
                               distinct_id                     start  \
0  antonietta.borzacchiello@deceuninck.com 2025-11-26 21:34:59+00:00   
1             cristobal.inostroza@otis.com 2025-11-25 13:56:27+00:00   
2                scarlett@blueimpresion.cl 2025-11-25 13:27:04+00:00   
3                scarlett@blueimpresion.cl 2025-11-25 16:22:16+00:00   

   marker_select_count  search_filter_select_count  page_view_count  \
0                    1                           1                3   
1                    0                           0                0   
2                    0                           1                5   
3                    3                           2                7   

   download_flag  real_duration  cluster  
0              0         49.750        2  
1              0          0.021        0  
2              0        412.374        2  
3              0        409.135        1  


## Analyze Cluster Characteristics

### Subtask:
Calculate and display the average value of each original (unscaled) metric for each of the 3 clusters. This table will help interpret the characteristics and behavior of users within each cluster.


**Reasoning**:
To understand the characteristics of each cluster, I will group `df_features` by the 'cluster' column and calculate the mean of the original numerical metrics for each cluster, then display the result.



In [ ]:
import pandas as pd

# Group df_features by 'cluster' and calculate the mean of the numerical columns
cluster_means = df_features.groupby('cluster')[numerical_cols].mean()

print("Average metrics for each cluster:")
print(cluster_means)

Average metrics for each cluster:
         marker_select_count  search_filter_select_count  page_view_count  \
cluster                                                                     
0                        0.0                         0.0              0.0   
1                        3.0                         2.0              7.0   
2                        0.5                         1.0              4.0   

         download_flag  real_duration  
cluster                                
0                  0.0          0.021  
1                  0.0        409.135  
2                  0.0        231.062  


## Final Task

### Subtask:
Summarize the clustering results and the insights gained from the average metrics per cluster.


## Summary:

### Data Analysis Key Findings

*   **Data Preparation:** Initial data loading and type conversion required careful handling of datetime formats and correct column identification. `df_events['timestamp']` was converted using `ISO8601` format, and `df_sessions['start']` and `df_sessions['end']` were converted using `%d-%m-%Y %H:%M:%S` format.
*   **Event-Session Association:** `df_events` and `df_sessions` were successfully merged based on `distinct_id`. A critical filtering step ensured that only events occurring within their respective session's start and end times were retained, resulting in 29 relevant event entries.
*   **Feature Engineering:** Five session-level metrics were calculated for 4 unique sessions:
    *   `marker_select_count`: Count of 'marker\_select' events.
    *   `search_filter_select_count`: Count of 'search' or 'filter\_select' events.
    *   `page_view_count`: Count of 'page\_view' events.
    *   `download_flag`: Binary indicator (1 if 'download' event occurred, 0 otherwise).
    *   `real_duration`: Session duration in seconds (difference between last and first event).
*   **Preprocessing for Clustering:** Missing values in numerical features were imputed with 0, and all numerical features were scaled using `StandardScaler` to prepare the data for clustering.
*   **KMeans Clustering:** The scaled data was clustered into 3 distinct groups using KMeans.
*   **Cluster Characteristics:** Analysis of average unscaled metrics for each cluster revealed distinct user behaviors:
    *   **Cluster 0 (Inactive/Short Sessions):** Characterized by very low activity across all metrics, with average `marker_select_count` and `search_filter_select_count` close to 0, and a `real_duration` also near 0.
    *   **Cluster 1 (Highly Engaged Users):** Shows the highest level of engagement with average values of 3.0 for `marker_select_count`, 2.0 for `search_filter_select_count`, 7.0 for `page_view_count`, and a substantial `real_duration` of approximately 409 seconds.
    *   **Cluster 2 (Moderately Engaged Users):** Exhibits moderate activity, with average `marker_select_count` of 0.5, `search_filter_select_count` of 1.0, `page_view_count` of 4.0, and a `real_duration` of approximately 231 seconds.

### Insights or Next Steps

*   The identified clusters represent different user engagement patterns, allowing for targeted strategies. For instance, Cluster 1 represents highly engaged users who could be ideal candidates for new feature adoption or loyalty programs, while Cluster 0 might need re-engagement campaigns.
*   Further investigation into the specific events and user journeys within each cluster, particularly Cluster 1 (highly engaged) and Cluster 0 (least engaged), could provide deeper insights into user motivations and pain points.
